core/destroy.go

```diff
@@ -1,15 +1,213 @@
 	"context"
+	"errors"
 	"fmt"
+	"os"
+	"path/filepath"
+	"strings"
+	"sync"
 
+	"github.com/charmbracelet/lipgloss"
 	"github.com/charmbracelet/log"
+	containerlablabels "github.com/srl-labs/containerlab/labels"
+	"github.com/srl-labs/containerlab/links"
+	"github.com/srl-labs/containerlab/nodes"
+	"github.com/srl-labs/containerlab/runtime"
 	"github.com/srl-labs/containerlab/runtime/ignite"
+	"github.com/srl-labs/containerlab/types"
+	"github.com/srl-labs/containerlab/utils"
 )
 
-// Destroy the given topology.
-func (c *CLab) Destroy(ctx context.Context, maxWorkers uint, keepMgmtNet bool) error {
+func (c *CLab) Destroy(ctx context.Context, options ...DestroyOption) (err error) {
+	opts := NewDestroyOptions()
+
+	for _, opt := range options {
+		opt(opts)
+	}
+
+	var containers []runtime.GenericContainer
+
+	if opts.all {
+		containers, err = c.ListContainers(ctx)
+	} else if c.TopoPaths.TopologyFilenameAbsPath() != "" {
+		containers, err = c.ListNodesContainersIgnoreNotFound(ctx)
+	} else {
+		var listOpts []ListOption
+
+		if c.Config.Name != "" {
+			listOpts = []ListOption{WithListLabName(c.Config.Name)}
+		} else {
+			listOpts = []ListOption{WithListContainerlabLabelExists()}
+		}
+
+		containers, err = c.ListContainers(ctx, listOpts...)
+	}
+
+	if err != nil {
+		return err
+	}
+
+	// topo will hold the reference to the topology file
+	// as the key and the respective lab directory as the referenced value
+	topos := map[string]string{}
+
+	for _, container := range containers {
+		topoFile, ok := container.Labels[containerlablabels.TopoFile]
+		if !ok {
+			continue
+		}
+
+		topos[topoFile] = filepath.Dir(container.Labels[containerlablabels.NodeLabDir])
+	}
+
+	defer func() {
+		err = c.destroyLabDirs(topos, opts.all)
+	}()
+
+	if len(topos) == 0 {
+		return nil
+	}
+
+	log.Debugf("got the following topologies for destroy: %+v", topos)
+
+	// if all, and cli doesnt have --yes flag, and in a terminal -- prompt user confirmation
+	if opts.all && opts.terminalPrompt && utils.IsTerminal(os.Stdin.Fd()) {
+		err := cliPromptToDestroyAll(topos)
+		if err != nil {
+			return err
+		}
+	}
+
+	var errs []error
+
+	for topo, labdir := range topos {
+		cc, err := c.cloneForDestroy(ctx, topo, labdir, *opts)
+		if err != nil {
+			log.Errorf("error creating clab instance for topo %q: %v", topo, err)
+
+			return err
+		}
+
+		err = cc.destroy(ctx, opts.maxWorkers, opts.keepMgmtNet)
+		if err != nil {
+			log.Errorf("Error occurred during the %s lab deletion: %v", cc.Config.Name, err)
+			errs = append(errs, err)
+		}
+	}
+
+	if len(errs) != 0 {
+		return fmt.Errorf("error(s) occurred during the deletion. Check log messages")
+	}
+
+	return err
+}
+
+// creates a mostly cloned version of the current c but set to the new topology, and with the
+// necessary steps (mgmt network things) handled preparing the new CLab for destruction.
+func (c *CLab) cloneForDestroy(ctx context.Context, topo, labDir string, opts DestroyOptions) (*CLab, error) {
```

Comment:
"clone" is perhaps poorly named -- basically just spawning a new clab instance for each topo like before. originally I was going to just copy things but runtime initialization is complicated and I revered to the normal creation process again